pkg/v1/cmd.go

```diff
@@ -0,0 +1,620 @@
+				// make up the pull request, but merge strategies other than "merge" will edit those commits - all the v1
+				// repos use such a strategy. We could reach out to the GitHub API, but that's expensive and time consming.
+
+				// Instead, we know that (today) all the v1 repos we're managing use a strategy that renames the commit
+				// to be sufffixed with the pull request "(#1234)", so we can check to see if such a commit exists upstream.
+				// This is not bullet-proof, but the failure mode is that we will try to cherry-pick a commit that does not
+				// need to be there and will either apply as a benign empty commit or fail to apply - both acceptable outcomes.
+				// We won't silently do the wrong thing.
+
+				rawMatches, err := internal.RunCommand(logger, internal.WithDir(exec.CommandContext(ctx,
+					"git", "log", "--pretty=format:%H", "--grep", fmt.Sprintf("(#%s)", match), commit,
+				), dir))
+				if err != nil {
+					return nil, err
+				}
+
+				if len(strings.TrimSpace(rawMatches)) == 0 {
+					logger.Info("cherry-picked PR needs to be carried")
+					downstreamCommits = append(downstreamCommits, info)
+				}
+			}
+		}
+	}
+	return downstreamCommits, nil
+}
+
+func applyConfig(ctx context.Context, logger *logrus.Entry, dir string, config Config, commitArgs []string) error {
+	// first, get us to the upstream target
+	for _, cmd := range [][]string{
+		{"git", "checkout", "main"},
+		{"git", "branch", "synchronize", "--force", config.Target.Hash},
+		{"git", "checkout", "synchronize"},
+		append([]string{"git", "merge", "--strategy", "ours", "main"}, commitArgs...),
+	} {
+		if _, err := internal.RunCommand(logger, internal.WithDir(exec.CommandContext(ctx,
+			cmd[0], cmd[1:]...,
+		), dir)); err != nil {
+			return err
+		}
+	}
+
+	// then, cherry-pick the additional bits
+	for _, commit := range config.Additional {
+		if _, err := internal.RunCommand(logger, internal.WithDir(exec.CommandContext(ctx,
+			"git", "cherry-pick", commit.Hash,
+		), dir)); err != nil {
+			return err
+		}
+	}
+
+	// finally, apply our generated patches on top
+	for _, cmd := range []*exec.Cmd{
+		internal.WithEnv(internal.WithDir(exec.CommandContext(ctx,
+			"go", "mod", "tidy",
+		), dir), os.Environ()...),
+		internal.WithEnv(internal.WithDir(exec.CommandContext(ctx,
+			"go", "mod", "vendor",
+		), dir), os.Environ()...),
+		internal.WithEnv(internal.WithDir(exec.CommandContext(ctx,
+			"go", "mod", "verify",
+		), dir), os.Environ()...),
+		internal.WithEnv(internal.WithDir(exec.CommandContext(ctx,
+			"go", "mod", "tidy",
+		), filepath.Join(dir, "openshift")), os.Environ()...),
+		internal.WithEnv(internal.WithDir(exec.CommandContext(ctx,
+			"go", "mod", "vendor",
+		), filepath.Join(dir, "openshift")), os.Environ()...),
+		internal.WithEnv(internal.WithDir(exec.CommandContext(ctx,
+			"go", "mod", "verify",
+		), filepath.Join(dir, "openshift")), os.Environ()...),
+		internal.WithDir(exec.CommandContext(ctx,
+			"git", "add", "--force",
```

Comment:
Technically this doesn't seem to do anything, because we do `git commit <specific files>` next. The docs say if you list specific files with `commit`, it ignores staged changes in the index.

Reply:
Good catch. I cargo-culted this from the v0 downstreaming shell script. Time to remove!